### PR TITLE
Add guest functionality to add customer command

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Adds ability to create guest customers to AddCustomerCommand. Added related tests that check that it works correctly. Extracted from https://github.com/PrestaShop/PrestaShop/pull/27769 to reduce it's scope.
| Type?             | refacto
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | Info below.
| Fixed ticket?     |
| Related PRs       |
| Sponsor company   |

### How to test
- The main thing to see us that tests are green. ✅
- This PR should have absolutely no effect on any backoffice work, but just to be sure - try to create a customer, try to edit him and see that it works fine.

### CODE REVIEW

File: src/Adapter/Customer/CommandHandler/AddCustomerHandler.php [hunk 1]
```diff
@@ -83,7 +84,13 @@ public function handle(AddCustomerCommand $command)
             throw new CustomerException('Customer contains invalid field values');
         }
 
-        $this->assertCustomerWithGivenEmailDoesNotExist($command->getEmail());
+        // If dealing with a registered customer, we need to check if the email does not exist.
+        // Two guests with the same email can co-exist, two registered customers can not.
+        if (!$command->isGuest()) {
```
Comment (Hlavtox) [inline src/Adapter/Customer/CommandHandler/AddCustomerHandler.php:89]: `$command->isGuest()` will always be false now, except in tests. No need to test anything manually.

File: src/Adapter/Customer/CommandHandler/AddCustomerHandler.php [hunk 2]
```diff
@@ -130,8 +140,9 @@ private function fillCustomerWithCommandData(Customer $customer, AddCustomerComm
         $customer->optin = $command->isPartnerOffersSubscribed();
         $customer->birthday = $command->getBirthday()->getValue();
         $customer->id_shop = $command->getShopId();
+        $customer->is_guest = $command->isGuest();
```
Comment (Hlavtox) [inline src/Adapter/Customer/CommandHandler/AddCustomerHandler.php:143]: `$customer->is_guest` is `false` by default, `$command->isGuest()` will always be false now, except in tests. No need to test anything manually.

File: src/Adapter/Customer/CommandHandler/AddCustomerHandler.php [hunk 3]
```diff
@@ -69,12 +69,13 @@ public function __construct(Hashing $hashing, $legacyCookieKey)
...
+        // Prepare new legacy customer object
         $customer = new Customer();
 
+        // Load it up with the data from the command
         $this->fillCustomerWithCommandData($customer, $command);
 
-        // validateFieldsRequiredDatabase() below is using $_POST
-        // to check if required fields are set
+        // validateFieldsRequiredDatabase() below is using $_POST to check if required fields are set
         $_POST[RequiredField::PARTNER_OFFERS] = $command->isPartnerOffersSubscribed();
```
Comment (zuk3975) [inline src/Adapter/Customer/CommandHandler/AddCustomerHandler.php:79]: CommandHandler MUST NOT be coupled with request. (also we never use $_POST or $_GET in symfony context, there is symfony Request object for cleaner way of getting request params).

Comment (zuk3975) [reply src/Adapter/Customer/CommandHandler/AddCustomerHandler.php:79]: With that said, this field must be probably added in your form, and passed in formDataHandler. FormData handler is responsible to handle the form data (request data) and convert it to related command.

Comment (Hlavtox) [reply src/Adapter/Customer/CommandHandler/AddCustomerHandler.php:79]: Julius, I did not add this code, we can do it in another PR. I don't to increase the scope million times again and never merge it.

File: tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php [hunk 4]
```diff
@@ -82,31 +84,46 @@ public function createACustomerUsingCommand($customerReference, TableNode $table
...
-            'password',
         ];
-
         foreach ($mandatoryFields as $mandatoryField) {
             if (!array_key_exists($mandatoryField, $data)) {
                 throw new Exception(sprintf('Mandatory property %s for customer has not been provided', $mandatoryField));
             }
         }
+        if (!array_key_exists('password', $data) && empty($data['isGuest'])) {
+            throw new Exception('Password must be provided, if creating a registered customer');
```
Comment (zuk3975) [inline tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php:99]: you must add this check in command and throw some CustomerConstraint exception if it fails. Then you are sure that its impossible to build "invalid command".
Then you add a test which expects that exception when you build the command without password and empty isGuest.

File: src/Core/Domain/Customer/Command/AddCustomerCommand.php [hunk 5]
```diff
@@ -394,4 +402,12 @@ public function setRiskId($riskId)
 
         return $this;
     }
+
+    /**
+     * @return bool
+     */
+    public function isGuest()
```
Comment (zuk3975) [inline src/Core/Domain/Customer/Command/AddCustomerCommand.php]:
```suggestion
    public function isGuest(): bool
```

File: src/Adapter/Customer/CommandHandler/AddCustomerHandler.php (hunk 3 above)
Comment (zuk3975) [reply src/Adapter/Customer/CommandHandler/AddCustomerHandler.php:79]: Oh right, I just saw it wasn't your change

File: tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php (hunk 4 above)
Comment (Hlavtox) [reply tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php:99]: What is the difference to the original `throw new Exception(sprintf('Mandatory property %s for customer has not been provided', $mandatoryField));`?

File: tests/Integration/Behaviour/Features/Scenario/Customer/customer_management.feature [hunk 6]
```diff
@@ -38,7 +38,22 @@ Feature: Customer Management
...
+    Then I should be returned an error message 'Registered customer with email "naapoler.dev@prestashop.com" already exists'
+
+  Scenario: Ability to create a guest customer, even if registered customer with that email exists
+    When I create a customer "CUST-3" with following properties:
+      | firstName | Mathieu                     |
+      | lastName  | Napoler                     |
+      | email     | naapoler.dev@prestashop.com |
+      | password  | PrestaShopForever1_!        |
+      | isGuest   | true                        |
+    When I query customer "CUST-3" I should get a Customer with properties:
```
Comment (zuk3975) [inline tests/Integration/Behaviour/Features/Scenario/Customer/customer_management.feature]: Optional and not important, just scenario reads better IMO
```suggestion
    And I query customer "CUST-3" I should get a Customer with properties:
```

File: tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php (hunk 4 above)
Comment (Hlavtox) [reply tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php:99]: These "required data" are only for test purposes.

File: tests/Integration/Behaviour/Features/Scenario/Customer/customer_management.feature (hunk 6 above)
Comment (Hlavtox) [reply tests/Integration/Behaviour/Features/Scenario/Customer/customer_management.feature]: Done ✅

File: src/Core/Domain/Customer/Command/AddCustomerCommand.php (hunk 5 above)
Comment (Hlavtox) [reply src/Core/Domain/Customer/Command/AddCustomerCommand.php]: Done ✅

File: tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php (hunk 4 above)
Comment (Hlavtox) [reply tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php:99]: I did not make the fields optional in the command in the end, these data are always required to construct the command.

Comment (zuk3975) [reply tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php:99]: oh right, the password field didn't change, it just got filled with random value in data handler :smile:

**Comment by zuk3975** [comment]
ping @PrestaShop/committers
1. I suggested @Hlavtox to move the logic from cqrs handler to data handler, but now Im not sure if it was right suggestion (its the quickest way of adding this feature though)

2. But handling the data depending if on $isGuest might have to be in the cqrs handler afterall. Then we would be able to test the command with behats and have less logic in the data handler, altough it would require introducing BC that way (because password field would become optional with the asserto depending of isGuest property)

3. Another solution would be to simply have a dedicated command for guest creation. Maybe part of current command could be extracted to some abstract class or other service to avoid code repetition. Though some code would be repeated, but at least the command would be foolproof, without bc breaks and tested by behats

Can anyone help which way is acceptable?

**Comment by Hlavtox** [comment]
- This is the easiest solution with no BC breaks.
- It solved my issue that I would need to do something with a Password object validation or make it optional.
- Also, maybe somebody would like to create a guest with different groups programatically.

If I don't do it, nobody will, and I want to continue working on the form.

File: src/Adapter/Customer/CommandHandler/AddCustomerHandler.php (hunk 1 above)
Comment (0x346e3730) [inline src/Adapter/Customer/CommandHandler/AddCustomerHandler.php:89]: What about a guest that uses the same email that a registered user ?

Comment (Hlavtox) [reply src/Adapter/Customer/CommandHandler/AddCustomerHandler.php:89]: That is allowed, thats why I put it inside the condition.

File: tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php [hunk 7]
```diff
@@ -82,31 +84,46 @@ public function createACustomerUsingCommand($customerReference, TableNode $table
...
+        }
+
+        // Process data and apply fallbacks
+        $password = $data['password'];
+        $defaultGroupId = isset($data['defaultGroupId']) ? $data['defaultGroupId'] : $defaultGroups->getCustomersGroup()->getId();
```
Comment (kpodemski) [inline tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php]:
```suggestion
        $defaultGroupId = $data['defaultGroupId'] ?? $defaultGroups->getCustomersGroup()->getId();
```

nitpick, I know, but to give you an example of how you could do things a bit cleaner 👍🏻

File: tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php [hunk 8]
```diff
@@ -82,31 +84,46 @@ public function createACustomerUsingCommand($customerReference, TableNode $table
...
+        $groupIds = isset($data['groupIds']) ? $data['groupIds'] : [$defaultGroups->getCustomersGroup()->getId()];
```
Comment (kpodemski) [inline tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php]:
```suggestion
        $groupIds = $data['groupIds'] ?? [$defaultGroups->getCustomersGroup()->getId()];
```

same

**Review by kpodemski — COMMENTED** [review]
Two very small comments, but in general, it looks legit to me. There are parts of these commands that could be improved, but this shouldn't impact your PR.

File: tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php (hunk 7 above)
Comment (Hlavtox) [reply tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php]: Honestly, I did not know how to use this syntax yet. I will keep it in mind 🚀

**Review by zuk3975 — APPROVED** [review]
I approve it. It probably can be done better (but hey, nothing is perfect, right? :smile: ). Anyway we can improve it later. At least it doesn't introduce a BC break now and enables Hlavtox working further on the feature.

**Review by kpodemski — APPROVED** [review]

**Review by djoelleuch — APPROVED** [review]
Hello @Hlavtox  ,

Automated tests :https://github.com/djoelleuch/ga.tests.ui.pr/actions/runs/4221365490/jobs/7329740039

LGTM ! It is QA :white_check_mark: 
Thanks!

**Comment by Hlavtox** [comment]
Awesome, thank you @djoelleuch! 🌹
